src/array/cpu/disjoint_union.cc

```diff
@@ -0,0 +1,126 @@
+/**
+*   Copyright (c) 2022, NVIDIA CORPORATION.
+*
+*   Licensed under the Apache License, Version 2.0 (the "License");
+*   you may not use this file except in compliance with the License.
+*   You may obtain a copy of the License at
+*
+*       http://www.apache.org/licenses/LICENSE-2.0
+*
+*   Unless required by applicable law or agreed to in writing, software
+*   distributed under the License is distributed on an "AS IS" BASIS,
+*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*   See the License for the specific language governing permissions and
+*   limitations under the License.
+*
+* \file array/cpu/disjoint_union.cc
+* \brief Disjoint union CPU implementation.
+*/
+
+#include <dgl/array.h>
+#include <dgl/runtime/parallel_for.h>
+#include <tuple>
+
+namespace dgl {
+using runtime::NDArray;
+namespace aten {
+namespace impl {
+
+template <DLDeviceType XPU, typename IdType>
+std::tuple<IdArray, IdArray, IdArray> _ComputePrefixSums(const std::vector<COOMatrix>& coos) {
+  IdArray prefix_src_arr = NewIdArray(
+    coos.size(), coos[0].row->ctx, coos[0].row->dtype.bits);
+  IdArray prefix_dst_arr = NewIdArray(
+    coos.size(), coos[0].row->ctx, coos[0].row->dtype.bits);
+  IdArray prefix_elm_arr = NewIdArray(
+    coos.size(), coos[0].row->ctx, coos[0].row->dtype.bits);
+
+  IdType* prefix_src = static_cast<IdType*>(prefix_src_arr->data);
+  IdType* prefix_dst = static_cast<IdType*>(prefix_dst_arr->data);
+  IdType* prefix_elm = static_cast<IdType*>(prefix_elm_arr->data);
+
+  dgl::runtime::parallel_for(0, coos.size(), [&](IdType b, IdType e){
+    for (IdType i = b; i < e; ++i) {
+      prefix_src[i] = coos[i].num_rows;
+      prefix_dst[i] = coos[i].num_cols;
+      prefix_elm[i] = coos[i].row->shape[0];
+    }
+  });
+
+  return std::make_tuple(CumSum(prefix_src_arr, true),
+                         CumSum(prefix_dst_arr, true),
+                         CumSum(prefix_elm_arr, true));
+}
+
+template <DLDeviceType XPU, typename IdType>
+COOMatrix DisjointUnionCoo(const std::vector<COOMatrix>& coos) {
+  bool has_data = false;
+  bool row_sorted = true;
+  bool col_sorted = true;
+  // check if data index array
+  for (size_t i = 0; i < coos.size(); ++i) {
+    CHECK_SAME_DTYPE(coos[0].row, coos[i].row);
+    CHECK_SAME_CONTEXT(coos[0].row, coos[i].row);
+    has_data |= COOHasData(coos[i]);
+  }
+
+  auto prefixes = _ComputePrefixSums<XPU, IdType>(coos);
+  auto prefix_src = static_cast<IdType*>(std::get<0>(prefixes)->data);
+  auto prefix_dst = static_cast<IdType*>(std::get<1>(prefixes)->data);
+  auto prefix_elm = static_cast<IdType*>(std::get<2>(prefixes)->data);
+
+  auto result_src = NewIdArray(
+    prefix_elm[coos.size()], coos[0].row->ctx, coos[0].row->dtype.bits);
+  auto result_dst = NewIdArray(
+    prefix_elm[coos.size()], coos[0].col->ctx, coos[0].col->dtype.bits);
+  auto result_dat = NullArray();
+  if (has_data) {
+    result_dat =  NewIdArray(
```

Comment:
extra space

Reply:
Probably I am missing; where do I allocate extra space?

Reply:
Maybe he means there is an extra space between "=" and "New" in line 78, a linting problem.

Reply:
I see :). Fixed.